src/tlo/dependencies.py

```diff
@@ -57,6 +57,69 @@ def get_all_dependencies(
     )
 
 
+def get_missing_dependencies(
+    module_instances: Iterable[Module],
+    get_dependencies: DependencyGetter = get_all_dependencies,
+) -> Set[str]:
+    """Get the set of missing required dependencies if any from an iterable of modules.
+
+    :param module_instances: Iterable of ``Module`` subclass instances to get missing
+        dependencies for.
+    :param get_dependencies: Callable which extracts the set of dependencies to check
+        for from a module instance. Defaults to extracting all dependencies.
+    :return: Set of ``Module`` subclass names corresponding to missing dependencies.
+    """
+    module_instances = list(module_instances)
+    modules_present = {type(module).__name__ for module in module_instances}
+    modules_present_are_alternatives_to = set.union(
+        # Force conversion to set to avoid errors when using set.union with frozenset
+        *(set(module.ALTERNATIVE_TO) for module in module_instances)
+    )
+    modules_required = set.union(
+        *(set(get_dependencies(module, modules_present)) for module in module_instances)
+    )
+
+    missing_dependencies = modules_required - modules_present
+    return (
+        missing_dependencies - modules_present_are_alternatives_to
+    )
+
+
+def initialise_missing_dependencies(modules, **module_kwargs):
+    """Get list of initialised instances of any missing dependencies for an iterable of modules.
+
+    :param modules: Iterable of ``Module`` subclass instances to get instances of missing
+        dependencies for.
+    :param module_kwargs: Any keyword arguments to use when initialising missing
+        module dependencies.
+    :return: List of ``Module`` subclass instances corresponding to missing dependencies.
+    """
+    module_instances = modules
+    module_class_map = get_module_class_map(set())
+    all_missing_dependencies = list()
+    all_module_instances = list(module_instances)
+
+    def get_all_missing_dependencies(module_instance: Iterable[Module]):
+        """ get all missing dependencies and return their module instances """
+        missing_dependencies = get_missing_dependencies(
+            module_instance, get_all_dependencies
+        )
+
+        if not missing_dependencies.issubset(all_missing_dependencies):
+            all_missing_dependencies.extend(missing_dependencies - set(all_missing_dependencies))
+            dependencies = [
+                module_class_map[dependency](**module_kwargs)
+                for dependency in missing_dependencies
+            ]
+            all_module_instances.extend(dependencies)
+            get_all_missing_dependencies(all_module_instances)
+
+        return all_module_instances
```

Comment:
As `all_module_instances` is defined outside this function, returning here is not needed an makes it difficult to track how variables are being assigned to. I would say it would be better to either not return anything and rely on `all_module_instances` already being available in the outer scope, or better still, removing the dependence of the `get_all_missing_dependencies` function on the `all_module_instances` variable in the outer scope and instead having it just depend on its arguments. As the `module_instances` variable passed as an argument to the call of the function below already has the same contents as `all_module_instances` (but is potentially not a list but another iterable type), this should be relatively simple to do.